### PR TITLE
Clean sluggification functions

### PR DESCRIPTION
**Pull request by jeanphilippegg**
This pull request just cleans up the default sluggification before exposing it
to users. The behavior is unchanged.

Here are the changes:

- Remove `denote-excluded-punctuation-regexp`. It was a `defconst`, so users
  could not modify it anyway. It also ended up being used "as-is" only for the
  title sluggification. We worked around it in both the signature and keyword
  sluggification. So it was not very useful in the end.
- Deprecate `denote-excluded-punctuation-extra-regexp`. Users should customize
  `denote-file-name-slug-functions`. It should be easy now just looking at the
  default sluggification.
- Removed the extra-characters parameter from `denote--slug-no-punct`. It was
  one of the workaround mentionned in the point above to have signatures
  remove hyphens as well.

The character regexp is slightly different for all sluggification function
(title, signature, keyword). That is why it is simply hard-coded in their
respective function. I think it also makes it easier to read for a user that
wants to customize it.

In this end, `denote--slug-no-punct` was just calling
`replace-regexp-in-string`, so I just removed it.

I have updated the manual and tests as well, but I recommend against dwelling
too much on the signature and keyword sluggification in the manual. The reason
is that I don't except users will feel the need to customize them at all.

Sluggification is necessary in the case of the title because we want to type
the title in the prompt as it will appear in the front matter and the
sluggification should clean it up for the file name. On the other hand, the
signature and the keywords can be typed correctly directly in the prompt.
There is not really a need to have a function clean them up for the file name.
Moreover, keywords (and signature probably when we include it in the front
matter) are sluggified in the front matter as well. Just mentionning
sluggification of signature/keyword as a possibility in the docstring of
`denote-file-name-slug-functions` seems enough to me.

### CODE REVIEW

**Comment by protesilaos** [comment]
Merged, thank you!

**Comment by protesilaos** [comment]
Do you think we need to include more examples for users in the manual? You know, the copy-paste kind of stuff which does work for a specific use-case.

**Comment by jeanphilippegg** [comment]
I think what we have in the manual is good!

We can close issue #328, keeping it as documentation here for those searching for this specific use-case.

**Comment by protesilaos** [comment]
> From: Jean-Philippe Gagné Guay ***@***.***>
> Date: Mon, 16 Sep 2024 16:12:00 -0700
>
> I think what we have in the manual is good!
>
> We can close issue #328, keeping it as documentation here for those
> searching for this specific use-case.

Very well!

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by protesilaos** [comment]
> We can close issue #328, keeping it as documentation here for those searching for this specific use-case.

I will write about it in the manual too, because the Info/HTML is easier to discover than coming to the issue tracker here.

File: README.org [hunk 1]
```diff
@@ -2271,24 +2273,20 @@ use the following in their Emacs configuration ([[#h:d375c6d2-92c7-425f-9d9d-219
...
+                              (denote-slug-keep-only-ascii str)))))
 
 (defun my-denote-sluggify-keyword (str)
   (downcase
-   (replace-regexp-in-string
-    "-" ""
-    (denote--slug-hyphenate
-     (denote--slug-no-punct
-      (denote-slug-keep-only-ascii str))))))
+   (replace-regexp-in-string "[][{}!@#$%^&*()+'\"?,.\|;:~`‘’“”/_ -=]*" ""
```
[the remark this answers is not in the document]
Comment (jeanphilippegg) [reply README.org:2288]: Oh, that is unintended! The issue is that the hyphen should be last in the character alternative (as the manual says).

You can replace the regexp with

```
"[][{}!@#$%^&*()+'\"?,.\|;:~`‘’“”/_ =-]*"
```

Thank you for reporting this!

@protesilaos Do you want me to make a pull request for this? We only need to change `denote-sluggify-keyword` (the other functions seem correct) and the relevant examples in the manual.

[the remark this answers is not in the document]
Comment (protesilaos) [reply README.org:2288]: Thank you @ssl19! @jeanphilippegg do whatever you prefer. I am fine, either way.

**Pull request merged**